### ISSUE TITLE
Non-exception based Result return value in C++

### ISSUE THREAD

**Issue by snowp** [dtolnay/cxx#1052]
The docs on returning a Result<T> from C++ seems to talk a bit about bridging with StatusOr<T> though doesn't provide any examples, so curious what the suggested way of doing this. I can imagine doing it by mapping a StatusOr to an exception on failure, then mapping the resulting Err into something more specific on the Rust side, or to return some non-Result type and then construct a Result<T> from that on the Rust side. Is this the best way to accomplish this? 

Given we don't have built in support for types like StatusOr, it would be nice to be able to somehow construct my own Result on the C++, but this doesn't seem possible from the docs and not sure how hard this would be to implement.

**Comment by Fulgen301** [comment]
This will probably become more relevant with C++23 and [std::expected](https://en.cppreference.com/w/cpp/header/expected).

**Comment by razaqq** [comment]
MS STL and libstdc++ now have support for std::expected, an option to translate Result<T> to that would be very helpful

**Comment by bridiver** [comment]
In the meantime what about a `rust::Result` cpp class for `RUST_CXX_NO_EXCEPTIONS`? Currently it looks like it calls `std::terminate` which is definitely not great and certainly doesn't match how Result is intended to work on the rust side.

**Comment by KitsuneRal** [comment]
Would it make sense, while the C++ ecosystem is catching up, to at least model `Result` as `std::variant` on the C++ side?

**Comment by razaqq** [comment]
no, all major standard library implementations (libc++, libstdc++, MS STL) now support `std::expected`

**Comment by HKalbasi** [comment]
If you are interested in having result in the C++ side, you may want to consider my library [Zngur](https://github.com/HKalbasi/zngur/). It has a different approach, and supports many Rust types without special support, including options and results. The [rustyline example](https://github.com/HKalbasi/zngur/blob/main/examples/rustyline/main.cpp) uses the rustyline crate in C++ without any glue code and handle its result using some ifs.
